File: libs/wasm-sdk/wasm-sdk-go/memory.go
package sdk

import "unsafe"

// packI64 packs a pointer and length into a single i64 value.
// Upper 32 bits = pointer, lower 32 bits = length.
func packI64(ptr uint32, length uint32) int64 {
	return int64(ptr)<<32 | int64(length)
}

// unpackI64 extracts pointer and length from a packed i64.
func unpackI64(packed int64) (ptr uint32, length uint32) {
	ptr = uint32(packed >> 32)
	length = uint32(packed & 0xFFFFFFFF)
	return
}

// stringToPtr returns the pointer and length for a Go string's underlying bytes.
func stringToPtr(s string) (uint32, uint32) {
	if len(s) == 0 {
		return 0, 0
	}
	b := []byte(s)
	return uint32(uintptr(unsafe.Pointer(&b[0]))), uint32(len(b))
}

// ptrToString reads a string from a wasm pointer and length.
func ptrToString(ptr uint32, length uint32) string {
	if ptr == 0 || length == 0 {
		return ""
	}
	b := make([]byte, length)
	src := unsafe.Pointer(uintptr(ptr))
	for i := uint32(0); i < length; i++ {
		b[i] = *(*byte)(unsafe.Pointer(uintptr(src) + uintptr(i)))
	}
	return string(b)
}

// unpackString reads a string from a packed i64 (ptr<<32|len).
func unpackString(packed int64) string {
	if packed == 0 {
		return ""
	}
	ptr, length := unpackI64(packed)
	return ptrToString(ptr, length)
}

// PackResult serializes a string to wasm memory and returns a packed i64.
// It keeps a reference to prevent GC.
var resultKeepAlive []byte

func PackResult(s string) int64 {
	b := []byte(s)
	resultKeepAlive = b
	if len(b) == 0 {
		return 0
	}
	ptr := uint32(uintptr(unsafe.Pointer(&b[0])))
	return packI64(ptr, uint32(len(b)))
}

// Alloc allocates a block of memory of the given size and returns a pointer.
//
//export alloc
func Alloc(size uint32) uint32 {
	buf := make([]byte, size)
	if size == 0 {
		return 0
	}
	return uint32(uintptr(unsafe.Pointer(&buf[0])))
}

// Dealloc is a no-op in Go (GC handles memory), but required by the host ABI.
//
//export dealloc
func Dealloc(ptr uint32, size uint32) {
}

// GetABIVersion returns the ABI version supported by this SDK.
//
//export get_abi_version
func GetABIVersion() int32 {
	return ABIVersion
}

File: libs/wasm-sdk/wasm-sdk-swift/Sources/FlowLikeSDK/Memory.swift
// Memory.swift — WASM memory management and ABI exports.

// MARK: - Pack / Unpack

/// Packs a pointer and length into a single i64: (ptr << 32) | len.
public func packI64(ptr: UInt32, len: UInt32) -> Int64 {
    (Int64(ptr) << 32) | Int64(len)
}

/// Unpacks a packed i64 into (ptr, len).
public func unpackI64(_ packed: Int64) -> (ptr: UInt32, len: UInt32) {
    let ptr = UInt32(truncatingIfNeeded: packed >> 32)
    let len = UInt32(truncatingIfNeeded: packed & 0xFFFF_FFFF)
    return (ptr, len)
}

// MARK: - String ↔ Pointer

/// Copies a Swift string into WASM linear memory and returns (ptr, len).
/// The caller is responsible for the allocated memory.
public func stringToPtr(_ s: String) -> (UInt32, UInt32) {
    let utf8 = Array(s.utf8)
    if utf8.isEmpty { return (0, 0) }
    let size = UInt32(utf8.count)
    let ptr = wasmAlloc(size)
    let dest = UnsafeMutablePointer<UInt8>(bitPattern: UInt(ptr))!
    for i in 0..<utf8.count {
        dest[i] = utf8[i]
    }
    return (ptr, size)
}

/// Reads a string from WASM linear memory at (ptr, len).
public func ptrToString(ptr: UInt32, len: UInt32) -> String {
    if ptr == 0 || len == 0 { return "" }
    let src = UnsafePointer<UInt8>(bitPattern: UInt(ptr))!
    let buf = UnsafeBufferPointer(start: src, count: Int(len))
    return String(decoding: buf, as: UTF8.self)
}

/// Reads a string from a packed i64.
public func unpackString(_ packed: Int64) -> String {
    if packed == 0 { return "" }
    let (ptr, len) = unpackI64(packed)
    return ptrToString(ptr: ptr, len: len)
}

// MARK: - Result packing

/// Serializes a string into WASM memory and returns a packed i64.
public func packResult(_ s: String) -> Int64 {
    let (ptr, len) = stringToPtr(s)
    return packI64(ptr: ptr, len: len)
}

/// Serializes a NodeDefinition to JSON and returns a packed i64.
public func serializeDefinition(_ def: NodeDefinition) -> Int64 {
    packResult(def.toJSON())
}

/// Serializes an ExecutionResult to JSON and returns a packed i64.
public func serializeResult(_ result: ExecutionResult) -> Int64 {
    packResult(result.toJSON())
}

// MARK: - ParseInput

/// Deserializes an ExecutionInput from WASM memory at (ptr, len).
public func parseInput(ptr: UInt32, length: UInt32) -> ExecutionInput {
    let json = ptrToString(ptr: ptr, len: length)
    return parseExecutionInput(json)
}

// MARK: - WASM ABI Exports

/// Allocates a block of memory in WASM linear memory.
@_cdecl("alloc")
public func wasmAlloc(_ size: UInt32) -> UInt32 {
    if size == 0 { return 0 }
    let ptr = UnsafeMutablePointer<UInt8>.allocate(capacity: Int(size))
    return UInt32(UInt(bitPattern: ptr))
}

/// Deallocates a block of memory in WASM linear memory.
@_cdecl("dealloc")
public func wasmDealloc(_ ptr: UInt32, _ size: UInt32) {
    guard ptr != 0, size != 0 else { return }
    let p = UnsafeMutablePointer<UInt8>(bitPattern: UInt(ptr))
    p?.deallocate()
}

/// Returns the ABI version supported by this SDK.
@_cdecl("get_abi_version")
public func getABIVersion() -> UInt32 {
    ABI_VERSION
}
